compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -2340,74 +2340,88 @@ private void analyzeExportableTypeRef(BSymbol owner, BTypeSymbol symbol, boolean
             return;
         }
         if (Symbols.isPublic(owner)) {
-            checkForExportableType(symbol, pos);
+            HashSet<BTypeSymbol> visitedSymbols = new HashSet<>();
+            checkForExportableType(symbol, pos, visitedSymbols);
         }
     }
 
-    private void checkForExportableType(BTypeSymbol symbol, Location pos) {
+    private void checkForExportableType(BTypeSymbol symbol, Location pos, HashSet<BTypeSymbol> visitedSymbols) {
 
         if (symbol == null || symbol.type == null || Symbols.isFlagOn(symbol.flags, Flags.TYPE_PARAM)) {
             // This is a built-in symbol or a type Param.
             return;
         }
+        if (visitedSymbols.contains(symbol)) {
+            return;
+        }
         switch (symbol.type.tag) {
             case TypeTags.ARRAY:
-                checkForExportableType(((BArrayType) symbol.type).eType.tsymbol, pos);
+                visitedSymbols.add(symbol);
```

Comment:
Why are we adding these here and not before the switch statement?

Reply:
Can also do 

```java
        if (!visitedSymbols.add(symbol)) {
            return;
        }
```

there instead.

Reply:
+1, no need of doing it separately. Will fix it